code/modules/mob/mob_helpers.dm

```diff
@@ -131,6 +131,41 @@ var/global/list/limb_types_by_name = list(
 		index++
 	return output_message
 
+// proc that parses an html input string and scrambles the non-html string contents
+/proc/stars_decode_html(message)
+	if(!length(message))
+		return
+
+	// todo: sanitize string more to remove unnencessary <>.
+	// list of parsed strings that have been scrambled, used to then reinsert the scrambled strings in to the message
+	var/list/replaced_scrambled_strings = list()
+
+	// boolean value to know if the current indexed element needs to be added to the scrambled message.
+	var/parsing_message = FALSE
+
+	// the current string value that needs to be scrambled.
+	var/current_string_to_scramble = ""
+
+	for(var/character_index in 1 to length(message))
+		if(message[character_index] == ">")
+			parsing_message = TRUE
+			continue
+		if(message[character_index] == "<")
+			parsing_message = FALSE
+			if(!length(current_string_to_scramble))
+				continue
+			var/scrambled_string = stars(current_string_to_scramble)
+			replaced_scrambled_strings[current_string_to_scramble] += scrambled_string
+			current_string_to_scramble = null
+			continue
+		if(parsing_message)
+			current_string_to_scramble += message[character_index]
+
+	for(var/unscrambled_message_key in replaced_scrambled_strings)
+		message = replacetext(message, unscrambled_message_key, replaced_scrambled_strings[unscrambled_message_key])
```

Comment:
Carefully-crafted user input would make this replace text inside of HTML tags. This should be reworked to be a bit more similar to how `/proc/stars` works, where the final message is being appended to in the original `for` loop. In this case, you could be appending `scrambled_string` from the loop before to a variable representing the final output. I suspect that would also be better performance-wise compared to repeatedly searching through the whole text for replacements.

Your current code also means that duplicates of `unscrambled_message_key` would get the same star text, while doing it iteratively would preserve the original behavior of not duplicating star text instances.

Reply:
I don't think the duplication matters too much, or the performance since this is very cold code. If a maintainer disagrees then I'll refactor it.